Wrong underlined argument in tuple returned by function

### Code

```Rust
fn get_arguments(data: &[u8]) -> (Vec<String>, &[u8]) {
    (Vec::new(), data)
}

let x = true;
let cmd = &[1, 2, 3, 4];
let (cmd, proc_args) = if x {
    get_arguments(cmd)
} else {
    (Vec::new(), &[])
};
```


### Current output

```Shell
error[E0277]: the size for values of type `[_]` cannot be known at compilation time
   --> src/unix/apple/macos/process.rs:581:10
    |
581 |     let (cmd, proc_args) = if x {
    |          ^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[_]`
    = note: all local variables must have a statically known size
    = help: unsized locals are gated as an unstable feature
```


### Desired output

```Shell
error[E0277]: the size for values of type `[_]` cannot be known at compilation time
   --> src/unix/apple/macos/process.rs:581:10
    |
581 |     let (cmd, proc_args) = if x {
    |               ^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[_]`
    = note: all local variables must have a statically known size
    = help: unsized locals are gated as an unstable feature
```


### Rationale and extra context

It's underlining the wrong argument.

### Other cases

_No response_

### Anything else?

_No response_

I'm getting an entirely different and much more helpful error:

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8ef43273677bf72e3c82fb026cf4d1b8

```
error[E0308]: `if` and `else` have incompatible types
  --> src/main.rs:11:5
   |
8  |   let (cmd, proc_args) = if x {
   |  ________________________-
9  | |     get_arguments(cmd)
   | |     ------------------ expected because of this
10 | | } else {
11 | |     (Vec::new(), &[])
   | |     ^^^^^^^^^^^^^^^^^ expected `(Vec<String>, &[u8])`, found `(Vec<_>, &[_; 0])`
12 | | };
   | |_- `if` and `else` have incompatible types
   |
   = note: expected tuple `(Vec<String>, &[u8])`
              found tuple `(Vec<_>, &[_; 0])`
```

Minimization:

```rust
fn issue_117846_repro() {
    let (cmd, _) = if true {
        produce() //
    } else {
        (Vec::new(), &[])
    };

    accept(&cmd);
}

fn produce() -> (Vec<String>, &'static [u8]) {
    todo!()
}

fn accept(cmd: &[String]) {}
```

Note that:

* in all cases (including when I ran the original code), there is now a more sensible type error, and
* the unsized local error is not actually pointing at the wrong thing, it's talking about the type of the part of the tuple unrelated to the if/else type mismatch.

It seems like the type mismatch is causing the compiler to fail to properly perform the deref coercion of the `Vec<String>`, even though it's not the part of the tuple that has the type mismatch.

```rust
error[E0308]: `if` and `else` have incompatible types
 --> src/lib.rs:5:9
  |
2 |       let (cmd, _) = if true {
  |  ____________________-
3 | |         produce() //
  | |         --------- expected because of this
4 | |     } else {
5 | |         (Vec::new(), &[])
  | |         ^^^^^^^^^^^^^^^^^ expected `(Vec<String>, &[u8])`, found `(Vec<_>, &[_; 0])`
6 | |     };
  | |_____- `if` and `else` have incompatible types
  |
  = note: expected tuple `(Vec<String>, &[u8])`
             found tuple `(Vec<_>, &[_; 0])`

error[E0277]: the size for values of type `[String]` cannot be known at compilation time
 --> src/lib.rs:2:10
  |
2 |     let (cmd, _) = if true {
  |          ^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `[String]`
  = note: all local variables must have a statically known size
  = help: unsized locals are gated as an unstable feature
```

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs
@@ -138,7 +138,20 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
     #[inline]
     pub fn write_ty(&self, id: hir::HirId, ty: Ty<'tcx>) {
         debug!("write_ty({:?}, {:?}) in fcx {}", id, self.resolve_vars_if_possible(ty), self.tag());
-        self.typeck_results.borrow_mut().node_types_mut().insert(id, ty);
+        let mut typeck = self.typeck_results.borrow_mut();
+        let mut node_ty = typeck.node_types_mut();
+        if let Some(ty) = node_ty.get(id)
+            && let Err(e) = ty.error_reported()
+        {
+            // Do not overwrite nodes that were already marked as `{type error}`. This allows us to
+            // silence unnecessary errors from obligations that were set earlier than a type error
+            // was produced, but that is overwritten by later analysis. This happens in particular
+            // for `Sized` obligations introduced in gather_locals. (#117846)
+            self.set_tainted_by_errors(e);
+            return;
+        }
+
+        node_ty.insert(id, ty);
 
         if let Err(e) = ty.error_reported() {
             self.set_tainted_by_errors(e);

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -1892,11 +1892,35 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         pat: &'tcx hir::Pat<'tcx>,
         ty: Ty<'tcx>,
     ) {
+        struct V<'tcx> {
+            tcx: TyCtxt<'tcx>,
+            pat_hir_ids: Vec<hir::HirId>,
+        }
+
+        impl<'tcx> Visitor<'tcx> for V<'tcx> {
+            type NestedFilter = rustc_middle::hir::nested_filter::All;
+
+            fn nested_visit_map(&mut self) -> Self::Map {
+                self.tcx.hir()
+            }
+
+            fn visit_pat(&mut self, p: &'tcx hir::Pat<'tcx>) {
+                self.pat_hir_ids.push(p.hir_id);
+                hir::intravisit::walk_pat(self, p);
+            }
+        }
         if let Err(guar) = ty.error_reported() {
             // Override the types everywhere with `err()` to avoid knock on errors.
             let err = Ty::new_error(self.tcx, guar);
             self.write_ty(hir_id, err);
             self.write_ty(pat.hir_id, err);
+            let mut visitor = V { tcx: self.tcx, pat_hir_ids: vec![] };
+            hir::intravisit::walk_pat(&mut visitor, pat);
+            // Mark all the subpatterns as `{type error}` as well. This allows errors for specific
+            // subpatterns to be silenced.
+            for hir_id in visitor.pat_hir_ids {
+                self.write_ty(hir_id, err);
+            }
             self.locals.borrow_mut().insert(hir_id, err);
             self.locals.borrow_mut().insert(pat.hir_id, err);
         }

--- a/compiler/rustc_middle/src/ty/typeck_results.rs
+++ b/compiler/rustc_middle/src/ty/typeck_results.rs
@@ -568,6 +568,11 @@ impl<'a, V> LocalTableInContextMut<'a, V> {
         self.data.get_mut(&id.local_id)
     }
 
+    pub fn get(&mut self, id: hir::HirId) -> Option<&V> {
+        validate_hir_id_for_typeck_results(self.hir_owner, id);
+        self.data.get(&id.local_id)
+    }
+
     pub fn entry(&mut self, id: hir::HirId) -> Entry<'_, hir::ItemLocalId, V> {
         validate_hir_id_for_typeck_results(self.hir_owner, id);
         self.data.entry(id.local_id)

--- a/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
+++ b/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs
@@ -2954,6 +2954,16 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                 }
             }
             ObligationCauseCode::VariableType(hir_id) => {
+                if let Some(typeck_results) = &self.typeck_results
+                    && let Some(ty) = typeck_results.node_type_opt(hir_id)
+                    && let ty::Error(_) = ty.kind()
+                {
+                    err.note(format!(
+                        "`{predicate}` isn't satisfied, but the type of this pattern is \
+                         `{{type error}}`",
+                    ));
+                    err.downgrade_to_delayed_bug();
+                }
                 match tcx.parent_hir_node(hir_id) {
                     Node::Local(hir::Local { ty: Some(ty), .. }) => {
                         err.span_suggestion_verbose(

--- a/tests/ui/sized/expr-type-error-plus-sized-obligation.rs
+++ b/tests/ui/sized/expr-type-error-plus-sized-obligation.rs
@@ -1,0 +1,22 @@
+#![allow(warnings)]
+
+fn issue_117846_repro() {
+    let (a, _) = if true { //~ ERROR E0277
+        produce()
+    } else {
+        (Vec::new(), &[]) //~ ERROR E0308
+    };
+
+    accept(&a);
+}
+
+struct Foo;
+struct Bar;
+
+fn produce() -> (Vec<Foo>, &'static [Bar]) {
+    todo!()
+}
+
+fn accept(c: &[Foo]) {}
+
+fn main() {}

--- a/tests/ui/sized/expr-type-error-plus-sized-obligation.rs
+++ b/tests/ui/sized/expr-type-error-plus-sized-obligation.rs
@@ -1,7 +1,7 @@
 #![allow(warnings)]
 
 fn issue_117846_repro() {
-    let (a, _) = if true { //~ ERROR E0277
+    let (a, _) = if true {
         produce()
     } else {
         (Vec::new(), &[]) //~ ERROR E0308

--- a/tests/ui/sized/expr-type-error-plus-sized-obligation.stderr
+++ b/tests/ui/sized/expr-type-error-plus-sized-obligation.stderr
@@ -14,17 +14,6 @@ LL | |     };
    = note: expected tuple `(Vec<Foo>, &[Bar])`
               found tuple `(Vec<_>, &[_; 0])`
 
-error[E0277]: the size for values of type `[Foo]` cannot be known at compilation time
-  --> $DIR/expr-type-error-plus-sized-obligation.rs:4:10
-   |
-LL |     let (a, _) = if true {
-   |          ^ doesn't have a size known at compile-time
-   |
-   = help: the trait `Sized` is not implemented for `[Foo]`
-   = note: all local variables must have a statically known size
-   = help: unsized locals are gated as an unstable feature
+error: aborting due to 1 previous error
 
-error: aborting due to 2 previous errors
-
-Some errors have detailed explanations: E0277, E0308.
-For more information about an error, try `rustc --explain E0277`.
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/sized/expr-type-error-plus-sized-obligation.stderr
+++ b/tests/ui/sized/expr-type-error-plus-sized-obligation.stderr
@@ -1,0 +1,30 @@
+error[E0308]: `if` and `else` have incompatible types
+  --> $DIR/expr-type-error-plus-sized-obligation.rs:7:9
+   |
+LL |       let (a, _) = if true {
+   |  __________________-
+LL | |         produce()
+   | |         --------- expected because of this
+LL | |     } else {
+LL | |         (Vec::new(), &[])
+   | |         ^^^^^^^^^^^^^^^^^ expected `(Vec<Foo>, &[Bar])`, found `(Vec<_>, &[_; 0])`
+LL | |     };
+   | |_____- `if` and `else` have incompatible types
+   |
+   = note: expected tuple `(Vec<Foo>, &[Bar])`
+              found tuple `(Vec<_>, &[_; 0])`
+
+error[E0277]: the size for values of type `[Foo]` cannot be known at compilation time
+  --> $DIR/expr-type-error-plus-sized-obligation.rs:4:10
+   |
+LL |     let (a, _) = if true {
+   |          ^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `[Foo]`
+   = note: all local variables must have a statically known size
+   = help: unsized locals are gated as an unstable feature
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0277, E0308.
+For more information about an error, try `rustc --explain E0277`.

